mlir/lib/Dialect/Tosa/IR/TosaCanonicalizations.cpp

```diff
@@ -73,56 +73,187 @@ struct SqrtReciprocalOptimization : public OpRewritePattern<tosa::PowOp> {
         rewriter.getFusedLoc({op.getLoc(), user->getLoc()}), outputType,
         op.getInput1());
     rewriter.replaceOp(user, rsqrtOp);
-      
+
     return success();
   }
 };
 
 void PowOp::getCanonicalizationPatterns(RewritePatternSet &results,
-                                           MLIRContext *context) {
+                                        MLIRContext *context) {
   results.add<SqrtReciprocalOptimization>(context);
 }
 
-LogicalResult SelectOp::canonicalize(SelectOp op, PatternRewriter &rewriter) {
-  auto notOp = op.getPred().getDefiningOp<tosa::LogicalNotOp>();
-  if (!notOp)
-    return failure();
-  rewriter.modifyOpInPlace(op, [&]() {
-    op.getOperation()->setOperands(
-        {notOp.getInput1(), op.getOnFalse(), op.getOnTrue()});
-  });
-  return success();
+struct SelectLogicalNotOptimization : public OpRewritePattern<tosa::SelectOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(tosa::SelectOp op,
+                                PatternRewriter &rewriter) const override {
+    auto notOp = op.getPred().getDefiningOp<tosa::LogicalNotOp>();
+    if (!notOp)
+      return failure();
+    rewriter.modifyOpInPlace(op, [&]() {
+      op.getOperation()->setOperands(
+          {notOp.getInput1(), op.getOnFalse(), op.getOnTrue()});
+    });
+    return success();
+  }
+};
+
+// This canonicalizes the following patterns:
+// %0 = tosa.greater_equal(input, x)
+// %1 = tosa.select(%0, input, x)
+// to tosa.clamp{min = x, max = max}(input)
+// and
+// %0 = tosa.greater_equal(input, x)
+// %1 = tosa.select(%0, x, input)
+// to tosa.clamp{min = min, max = x}(input)
+// The first pattern occurs in decompositions of LeakyReLU/PReLU with an alpha
+// of zero
+struct SelectToClampOptimization : public OpRewritePattern<tosa::SelectOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(tosa::SelectOp op,
+                                PatternRewriter &rewriter) const override {
+
+    auto geq = op.getPred().getDefiningOp<tosa::GreaterEqualOp>();
+    if (!geq) {
+      return rewriter.notifyMatchFailure(op,
+                                         "Predicate is not a GreaterEqualOp");
+    }
+
+    DenseElementsAttr geqIn2Attr;
+    if (!matchPattern(geq.getInput2(), m_Constant(&geqIn2Attr))) {
+      return rewriter.notifyMatchFailure(
+          op, "RHS of predicate GreaterEqualOp is not a constant");
+    }
+    auto isCompatibleSplat = [](DenseElementsAttr a,
+                                DenseElementsAttr b) -> bool {
+      if (!a.isSplat() || !b.isSplat()) {
+        return false;
+      }
+      if (llvm::isa<IntegerType>(a.getElementType())) {
+        return a.getSplatValue<APInt>() == b.getSplatValue<APInt>();
+      }
+      if (llvm::isa<FloatType>(a.getElementType())) {
+        return a.getSplatValue<APFloat>() == b.getSplatValue<APFloat>();
+      }
+      return false; // Only int and float types are supported
```

Comment:
Can you add a test cases that triggers this kind of scenario where we do not have splats